### PR TITLE
[Merged by Bors] - Minimum Outbound-Only Peers Requirement

### PR DESCRIPTION
**Pull request by kevlu93**
## Issue Addressed

#2325 

## Proposed Changes

This pull request changes the behavior of the Peer Manager by including a minimum outbound-only peers requirement. The peer manager will continue querying for peers if this outbound-only target number hasn't been met. Additionally, when peers are being removed, an outbound-only peer will not be disconnected if doing so brings us below the minimum. 

## Additional Info

Unit test for heartbeat function tests that disconnection behavior is correct. Continual querying for peers if outbound-only hasn't been met is not directly tested, but indirectly through unit testing of the helper function that counts the number of outbound-only peers.

EDIT: Am concerned about the behavior of ```update_peer_scores```. If we have connected to a peer with a score below the disconnection threshold (-20), then its connection status will remain connected, while its score state will change to disconnected. 

```rust
let previous_state = info.score_state();            
// Update scores            
info.score_update();
Self::handle_score_transitions(                
               previous_state,
                peer_id,
                info, 
               &mut to_ban_peers,
               &mut to_unban_peers,
               &mut self.events,
               &self.log,
);
```

```previous_state``` will be set to Disconnected, and then because ```handle_score_transitions``` only changes connection status for a peer if the state changed, the peer remains connected. Then in the heartbeat code, because we only disconnect healthy peers if we have too many peers, these peers don't get disconnected. I'm not sure realistically how often this scenario would occur, but it might be better to adjust the logic to account for scenarios where the score state implies a connection status different from the current connection status.

### CODE REVIEW

**Comment by paulhauner** [comment]
Thanks @kevlu93! This would be of interest to @AgeManning and @divagant-martian.

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs [hunk 1]
```diff
@@ -940,19 +947,31 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
...
-                //we only need to disconnect peers with healthy scores, since the others got already
-                //disconnected in update_peer_scores
-                .filter(|(_, info)| info.score_state() == ScoreState::Healthy)
             {
-                disconnecting_peers.push(**peer_id);
+                if peers_removed == connected_peer_count - self.target_peers {
+                    break;
+                }
+                if info.is_outbound_only() {
+                    if n_outbound_removed < outbound_only_peer_count - min_outbound_only_target {
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: Looks to me like this can underflow if 
`connected_peer_count  > target_peers` but `outbound_only_peer_count` > `min_outbound_only_target` 

Probably safer to use safe math everywhere and log the errors. We'd probably want to not do this calculation if this is the case tho, so make it conditional that `outbound_only_peer_count > min_outbound_only_target`

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs [hunk 2]
```diff
@@ -940,19 +947,31 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
...
+                        n_outbound_removed += 1;
+                    } else {
+                        continue;
+                    }
+                }
+                peers_removed += 1;
+                if info.score_state() == ScoreState::Healthy {
+                    disconnecting_peers.push(**peer_id);
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: Seems here that we only disconnect peers that are in the healthy state (this is the correct logic) but you are incrementing the removed count outside of this statement, similarly with `n_outbound_remove`. The result looks like we might not actually disconnect as many nodes as we think we are.

Also, you could just use `disconnecting_peers.len()` rather than tracking the separate variable `peers_removed` to enforce consistency.

**Review by AgeManning — CHANGES_REQUESTED** [review]
Your tests look good. 

It might be good to add to your tests to check the edge cases that I've pointed out (to make sure I'm correct :) ).

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs (hunk 1 above)
Comment (divagant-martian) [inline beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: `outbound_only_peer_count` has no guarantee to be greater than `min_outbound_only_target`

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs [hunk 3]
```diff
@@ -940,19 +947,31 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
...
-                //disconnected in update_peer_scores
-                .filter(|(_, info)| info.score_state() == ScoreState::Healthy)
             {
-                disconnecting_peers.push(**peer_id);
+                if peers_removed == connected_peer_count - self.target_peers {
+                    break;
+                }
+                if info.is_outbound_only() {
+                    if n_outbound_removed < outbound_only_peer_count - min_outbound_only_target {
+                        n_outbound_removed += 1;
```
Comment (divagant-martian) [inline beacon_node/eth2_libp2p/src/peer_manager/mod.rs:965]: you count this guy as removed, but there is one extra check a couple of lines after that would alter this count

**Review by divagant-martian — CHANGES_REQUESTED** [review]
I asked about @age about the purpose of the issue, so I'm just pointing some logic that might need some checks

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs (hunk 1 above)
Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: I might be interpreting you incorrectly, but if we use that as the conditional, isn't it possible that we could have removed some "bad" outbound-only peers, but do not because the conditional evaluates as false and then we iterate to the next. ie. in the test case I made, if we had another outbound-only peer that also had a score of -1, the two outbound-only peers will have the worst scores, but because ```outbound_only_peer_count > min_outbound_only_target```, we end up removing 2 peers with better scores but are not outbound-only. With the current logic it would instead remove one outbound-only peer and one other peer, so if an outbound-only peer has a bad score we would be able to remove it, while still meeting the quota.

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs (hunk 2 above)
Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: I think this behavior is meant to mimic the logic of the original code, which took ```connected_peer_count - self.target_peers```  peers before filtering out the ones that are not healthy. This is also why I used ```peers_removed``` instead of ```disconnecting_peers.len()```

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs (hunk 3 above)
Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs:965]: I meant to mimic the behavior of the original code, which kept ```connected_peer_count - self.target_peers``` peers and then further filtered the list based on whether they were healthy or not. Since ```update_peer_scores``` would have already disconnected the unhleathy peers, it should increment the count, even if we don't add it to the disconnecting_peers list.

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs (hunk 1 above)
Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: is this an issue tho? the conditional would evaluate to false, and we would not disconnect the outbound-only peer, which is the intended behavior

Comment (divagant-martian) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: It is an issue. For one logic should not depend on overflowing. This line with the right conditions panics on debug an evaluates to `true`pretty much always in release (0-1 in release is 18446744073709551615, the max value for usize)
I'll check your code again to give a better advice on how to accomplish what you want

Comment (AgeManning) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: Yes. I think you may have missed mine and Diva's point with our comments.
In rust, this statement: `outbound_only_peer_count - min_outbound_only_target ` will panic if `outbound_only_peer_count` is < `min_outbound_only_target` and will crash the lighthouse client (if run in debug :p). 

I'm suggesting to not do this subtraction in this case to avoid crashing the client. Rust has some SafeMath alternatives such as `outbound_only_peer_count.saturating_sub(min_outbound_only_target)` which won't panic but will saturate to 0.

Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs:965]: Ah good point! If I change the logic to ```min_outbound_only_target < n_outbound_removed + outbound_only_peer_count``` this would then handle the overflow issue while keeping the logic right?

Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: Yup just realized what you guys meant. Thanks for pointing it out! Was caught up in thinking through the logic that I didn't consider that these are unsigned!

Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: Would changing the conditional to ```min_outbound_only_target < outbound_only_peer_count - n_outbound_removed``` preserve the logic and handle the overflow issue? ```n_outbound_removed``` is bounded above by ```outbound_only_peer_count``` since it can only be incremented when it's looking at an outbound-only peer, so then we should be able to avoid the issue without needing to use SafeMath?

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs (hunk 2 above)
Comment (AgeManning) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: Oh yeah. Looks to me like its a bug in the old impl. I think the old logic is probably wrong here. Nice catch

Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: I actually think the old impl might be fine? Since as the comments mention, update_peer_scores will have disconnected the unhealthy ones already, so it's fine to not add them to the disconnecting_peers list since it's redundant. Or I might be missing something. It def tripped me up looking at the ordering of the filters

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs (hunk 1 above)
Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: Regardless I will add some more test cases since yea that would have been a pretty big issue

Comment (AgeManning) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: Yep, this seems like a sensible solution :).

Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: see my comment above to Age, think changing the conditional to ```min_outbound_only_target < outbound_only_peer_count - n_outbound_removed``` should preserve the logic while preventing the panic

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs (hunk 2 above)
Comment (AgeManning) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: Ah yes. Sorry, its been a while since I was in here and this has more complex logic than I recall. 

Yeah, looks like the old logic is correct, for peers that have transitioned into a disconnected or worse state we should try and disconnect from them. They could appear here if they are still connected and we want to avoid sending a double disconnection message. 

Your logic here looks correct, sorry I only did a quick skim of the PR :p.

Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs]: No worries! You guys got a lot of work goin on, just glad I wasn't totally off with my implementation

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs [hunk 4]
```diff
@@ -940,19 +946,33 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
         let connected_peer_count = self.network_globals.connected_peers();
         if connected_peer_count > self.target_peers {
             //remove excess peers with the worst scores, but keep subnet peers
-            for (peer_id, _) in self
+            //must also ensure that the outbound-only peer count does not go below the minimum threshold
```
Comment (divagant-martian) [inline beacon_node/eth2_libp2p/src/peer_manager/mod.rs:949]: We should have mentioned this earlier, sorry about that. We try to maintain a comment style of Uppercase first letter, space after `//`and final point. It's a bit recent that we formalized this [here](https://lighthouse-book.sigmaprime.io/contributing.html#comments) which is probably why the comment on the previous line to this one does not fit the desired style either. Would you mind updating that one too?

**Review by divagant-martian — CHANGES_REQUESTED** [review]
Leaving as a comment here, to avoid polluting your PR. We should have mentioned this earlier, so sorry about that. 

We try to maintain a comment style of Uppercase first letter, space after `//` and final point. It's a bit recent that we formalized this [here](https://lighthouse-book.sigmaprime.io/contributing.html#comments) which is probably why comments from the code you are improving do not fit this style either

Could you update yours to reflect this style?
I will give the logic another view and give a review later

Again thanks for your contribution and welcome to lighthouse

**Comment by AgeManning** [comment]
I'm going to leave this guy in @divagant-martian's capable hands.

File: beacon_node/eth2_libp2p/src/peer_manager/score.rs [hunk 5]
```diff
@@ -216,6 +216,13 @@ impl RealScore {
         self.set_lighthouse_score(0f64);
     }
 
+    #[cfg(test)]
+    // Set the gossipsub_score to a specific f64.
+    // Used in testing to induce score status changes during a heartbeat.
```
Comment (divagant-martian) [inline beacon_node/eth2_libp2p/src/peer_manager/score.rs]: Minor style nitpick: We prefer comments after attributes (same guideline I linked before) Also, would you please make this one a doc comment? Thanks

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs [hunk 6]
```diff
@@ -939,20 +945,34 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
 
         let connected_peer_count = self.network_globals.connected_peers();
         if connected_peer_count > self.target_peers {
-            //remove excess peers with the worst scores, but keep subnet peers
-            for (peer_id, _) in self
+            // Remove excess peers with the worst scores, but keep subnet peers.
```
Comment (divagant-martian) [inline beacon_node/eth2_libp2p/src/peer_manager/mod.rs:948]: After some bisect digging, this code was originally added in #1383. By that time we were not handling the intermediate "Disconnecting" peers. This was later added in #1796. I think this additional logic to account for is no longer needed and as noted by all, it hinders readability. 
If a peer got to this list (`worst_connected_peers`) it means it is really connected.

I'd prefer if you try to remove the "is healthy logic" and check if it works as expected.

Comment (divagant-martian) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs:948]: I ran your tests removing the is healthy logic and they all passed, and the tests look good, so I think this is Ok, but give it a peruse anyway. If you think I'm right, this means update a pair of comments in the tests that mention this logic. If I'm not, I think the `peers_removed += 1;` deserves it's own comment explaining logic for the future. Otherwise I think your PR is good to go

**Review by divagant-martian — CHANGES_REQUESTED** [review]
Looks good. Check comments about the "is healthy" logic. Marked as changes requested for minor nitpick. Let me know what you think and we are good to go

**Comment by divagant-martian** [comment]
About your concern, I think it's fine. If we were to disconnect peers in the disconnected score state it would be the same as banning them

File: beacon_node/eth2_libp2p/src/peer_manager/mod.rs (hunk 6 above)
Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs:948]: Okay will take a look! That'd be good if we can remove the condition since yea it is quite confusing.

Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs:948]: I think you're right, the "is healthy logic" is unnecessary. Like you said, the ```worst_connected_peers``` only has peers that are still connected. Thanks for the catch! I'll update to remove references to that logic.

Comment (kevlu93) [reply beacon_node/eth2_libp2p/src/peer_manager/mod.rs:948]: We can then use a loop with ```disconnecting_peers.len()``` that I think you / Age suggested earlier.

**Review by AgeManning — APPROVED** [review]
Looks good to @divagant-martian , looks good to me :)

**Comment by paulhauner** [comment]
bors -

**Comment by paulhauner** [comment]
bors r+

:crossed_fingers:
